internal/local/command_kafka_start.go

```diff
@@ -230,60 +268,90 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
+		} else {
+			if err := plaintextLn.Close(); err != nil {
+				return err
+			}
 		}
 	}
 
 	return nil
 }
 
-func getContainerEnvironmentWithPorts(ports *config.LocalPorts) []string {
-	return []string{
-		"KAFKA_BROKER_ID=1",
+func getContainerEnvironmentWithPorts(ports *config.LocalPorts, idx int32, numBrokers int32) []string {
+	brokerId := idx + 1
+	envs := []string{
+		fmt.Sprintf("KAFKA_BROKER_ID=%d", brokerId),
 		"KAFKA_LISTENER_SECURITY_PROTOCOL_MAP=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT",
-		fmt.Sprintf("KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://broker:%s,PLAINTEXT_HOST://localhost:%s", ports.BrokerPort, ports.PlaintextPort),
+		fmt.Sprintf("KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://confluent-local-broker-%d:%s,PLAINTEXT_HOST://localhost:%s", brokerId, ports.BrokerPorts[idx], ports.PlaintextPorts[idx]),
 		"KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR=1",
 		"KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS=0",
 		"KAFKA_TRANSACTION_STATE_LOG_MIN_ISR=1",
 		"KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR=1",
 		"KAFKA_PROCESS_ROLES=broker,controller",
-		"KAFKA_NODE_ID=1",
-		fmt.Sprintf("KAFKA_CONTROLLER_QUORUM_VOTERS=1@broker:%s", ports.ControllerPort),
-		fmt.Sprintf("KAFKA_LISTENERS=PLAINTEXT://broker:%s,CONTROLLER://broker:%s,PLAINTEXT_HOST://0.0.0.0:%s", ports.BrokerPort, ports.ControllerPort, ports.PlaintextPort),
+		fmt.Sprintf("KAFKA_NODE_ID=%d", brokerId),
+		getKafkaControllerQuorumVoters(ports, numBrokers),
+		fmt.Sprintf("KAFKA_LISTENERS=PLAINTEXT://confluent-local-broker-%d:%s,CONTROLLER://confluent-local-broker-%d:%s,PLAINTEXT_HOST://0.0.0.0:%s", brokerId, ports.BrokerPorts[idx], brokerId, ports.ControllerPorts[idx], ports.PlaintextPorts[idx]),
 		"KAFKA_INTER_BROKER_LISTENER_NAME=PLAINTEXT",
 		"KAFKA_CONTROLLER_LISTENER_NAMES=CONTROLLER",
 		"KAFKA_LOG_DIRS=/tmp/kraft-combined-logs",
 		"KAFKA_REST_HOST_NAME=rest-proxy",
-		fmt.Sprintf("KAFKA_REST_BOOTSTRAP_SERVERS=broker:%s", ports.BrokerPort),
-		fmt.Sprintf("KAFKA_REST_LISTENERS=http://0.0.0.0:%s", ports.KafkaRestPort),
 	}
+	if idx == 0 { // configure Kafka REST proxy broker 1
+		envs = append(envs, fmt.Sprintf("KAFKA_REST_LISTENERS=http://0.0.0.0:%s", ports.KafkaRestPort))
+		envs = append(envs, getKafkaRestBootstrapServers(ports, numBrokers))
+	}
+	return envs
+}
+
+func getNatPlaintextPorts(ports *config.LocalPorts, numBrokers int32) []nat.Port {
+	res := []nat.Port{}
+	for i := 0; i < int(numBrokers); i++ {
+		res = append(res, nat.Port(ports.PlaintextPorts[i]+"/tcp"))
+	}
+	return res
+}
+
+func getKafkaControllerQuorumVoters(ports *config.LocalPorts, numBrokers int32) string {
+	voters := fmt.Sprintf("KAFKA_CONTROLLER_QUORUM_VOTERS=1@confluent-local-broker-1:%s", ports.ControllerPorts[0])
+	for i := int32(1); i < numBrokers; i++ {
+		voters += fmt.Sprintf(",%d@confluent-local-broker-%d:%s", i+1, i+1, ports.ControllerPorts[i])
+	}
+	return voters
+}
+
+func getKafkaRestBootstrapServers(ports *config.LocalPorts, numBrokers int32) string {
```

Comment:
Seems like we could generalize this function and the one above?

Reply:
they're using different ports and different ways to construct the string. it's clearer to keep them separate.